community/codegen/src/main/java/org/neo4j/codegen/bytecode/ByteCodeExpressionVisitor.java

```diff
@@ -0,0 +1,600 @@
+    {
+        target.accept( this );
+        methodVisitor
+                .visitFieldInsn( GETFIELD, byteCodeName( field.owner() ), field.name(), typeName( field.type() ) );
+}
+
+    @Override
+    public void constant( Object value )
+    {
+        if ( value == null )
+        {
+            methodVisitor.visitInsn( ACONST_NULL );
+        }
+        else if ( value instanceof Integer )
+        {
+            pushInteger( (Integer) value );
+        }
+        else if ( value instanceof Byte )
+        {
+            pushInteger( (Byte) value );
+        }
+        else if ( value instanceof Short )
+        {
+            pushInteger( (Short) value );
+        }
+        else if ( value instanceof Long )
+        {
+            methodVisitor.visitLdcInsn( value );
+        }
+        else if ( value instanceof Double )
+        {
+            methodVisitor.visitLdcInsn( value );
+        }
+        else if ( value instanceof Float )
+        {
+            methodVisitor.visitLdcInsn( value );
+        }
+        else if ( value instanceof Boolean )
+        {
+            boolean b = (boolean) value;
+            methodVisitor.visitInsn( b ? ICONST_1 : ICONST_0 );
+        }
+        else
+        {
+            methodVisitor.visitLdcInsn( value );
+        }
+    }
+
+    @Override
+    public void getStatic( FieldReference field )
+    {
+        methodVisitor
+                .visitFieldInsn( GETSTATIC, byteCodeName( field.owner() ), field.name(), typeName( field.type() ) );
+    }
+
+    @Override
+    public void loadThis( String sourceName )
+    {
+        methodVisitor.visitVarInsn( ALOAD, 0 );
+    }
+
+    @Override
+    public void newInstance( TypeReference type )
+    {
+        methodVisitor.visitTypeInsn( NEW, byteCodeName( type ) );
+        methodVisitor.visitInsn( DUP );
+    }
+
+    @Override
+    public void not( Expression expression )
+    {
+        expression.accept( this );
+        Label l0 = new Label();
+        methodVisitor.visitJumpInsn( IFNE, l0 );
+        methodVisitor.visitInsn( ICONST_1 );
+        Label l1 = new Label();
+        methodVisitor.visitJumpInsn( GOTO, l1 );
+        methodVisitor.visitLabel( l0 );
+        methodVisitor.visitInsn( ICONST_0 );
+        methodVisitor.visitLabel( l1 );
+    }
+
+    @Override
+    public void ternary( Expression test, Expression onTrue, Expression onFalse )
+    {
+        ternaryExpression( IFEQ, test, onTrue, onFalse );
+    }
+
+    @Override
+    public void ternaryOnNull( Expression test, Expression onTrue, Expression onFalse )
+    {
+       ternaryExpression( IFNONNULL, test, onTrue, onFalse );
+    }
+
+    @Override
+    public void ternaryOnNonNull( Expression test, Expression onTrue, Expression onFalse )
+    {
+        ternaryExpression( IFNULL, test, onTrue, onFalse );
+    }
+
+    @Override
+    public void equal( Expression lhs, Expression rhs, TypeReference type )
+    {
+        switch ( type.simpleName() )
+        {
```

Comment:
Have we completely removed the notion of `Expression` carrying a type? It seems to me that the type parameter is a bit superfluous.